idaes/dmf/getver.py

```diff
@@ -0,0 +1,226 @@
+
+
+class GetVersionError(Exception):
+    pass
+
+
+class ModuleImportError(GetVersionError):
+    pass
+
+
+class PipError(GetVersionError):
+    pass
+
+
+class GitHashError(GetVersionError):
+    pass
+
+
+# Classes and functions
+
+
+#: Return from `Versioned.get_info()`
+VersionInfo = namedtuple("VersionInfo", "version git_hash")
+
+
+class Versioned:
+    """Get version information for a module.
+
+    Sample usage::
+
+        from idaes.dmf.getver import Versioned
+        import some_module
+
+        vrs = Versioned(some_module)
+        print(vrs.version)
+        print(vrs.git_hash)
+
+    """
+
+    #: PIP command
+    PIP = "pip"
+
+    def __init__(self, obj: Union[ModuleType, str]):
+        """Constructor.
+
+        Args:
+            obj: Imported package or package name.
+        """
+        if hasattr(obj, "__package__"):
+            mod = obj
+        else:
+            try:
+                mod = importlib.import_module(obj)
+            except ModuleNotFoundError as err:
+                raise ModuleImportError(f"Could not import module '{obj}': {err}")
+        self._mod = mod
+        self._name = mod.__package__
+        self._version = self._get_version()
+        if self._version is None:
+            raise PipError(f"Could not get version from pip for '{self._name}'")
+        self._hash = None
+
+    def get_info(self) -> VersionInfo:
+        """Get hash and version without throwing any exceptions.
+        If there was an exception, the hash may be None.
+
+        Returns:
+            The Git hash (which may be None) and the version (a string)
+        """
+        hash = None
+        try:
+            hash = self.git_hash
+        except GetVersionError:
+            pass
+        return VersionInfo(git_hash=hash, version=self.version)
+
+    @property
+    def name(self) -> str:
+        """Get the name of the module.
+
+        Returns:
+            Module name.
+        """
+        return self._name
+
+    @property
+    def version(self) -> str:
+        """Get the version.
+
+        Return:
+            The version as a string, e.g. "2.1.0"
+        """
+        return self._version
+
+    @property
+    def git_hash(self) -> str:
+        """Get the git hash.
+
+        Returns:
+            Git hash as a string of hexadecimal digits, e.g. "a5327b889b4606792cd31194f939c0d358e68e43"
+
+        Raises:
+            GitHashError
+        """
+
+        if self._hash is None:
+            self._hash = self._get_git_hash()
+        return self._hash
+
+    # -- Protected methods -- #
+
+    def _get_version(self):
+        status, output = subprocess.getstatusoutput(f"{self.PIP} list")
+        if status != 0:
+            raise PipError(f"Shell command '{self.PIP} list' failed")
+        version = None
+        for s in output.split("\n"):
+            if s.startswith(self._name):
+                version = s.split()[1]
+                break
+        return version
+
+    def _get_git_hash(self):
```

Comment:
One question, instead of adding another dependency for IDAES, would it be sufficient to just do something like:
```python
path = self._get_module_path()
while not os.path.exists(os.path.join(path, '.git')):
    tmp = os.path.split(path)[0]
    if tmp == path:
        return None
    path = tmp
with open(os.path.join(path, '.git', 'ORIG_HEAD'), 'r') as FILE:
    hash = FILE.readline().strip()
return hash
```

Reply:
Assuming that git doesn't change things, I suppose so.